### PR TITLE
[RLlib] "Light-weighted" RLModules for use in inference-only or exploration-only settings.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

`RLModule`'s used in inference or exploration do not need critic or target networks. Without them they are "light-weighted" and synching should be faster b/c less parameters have to be synched. 

This PR proposes a solution to synch "Light-weighted" modules. 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/dqn/torch/dqn_rainbow_torch_rl_module.py [hunk 1]
```diff
@@ -37,17 +37,18 @@ def setup(self):
         # If we use a noisy encoder. Note, only if the observation
         # space is a flat space we can use a noisy encoder.
         self.uses_noisy_encoder = isinstance(self.encoder, TorchNoisyMLPEncoder)
-        # We do not want to train the target networks.
-        # AND sync all target nets with the actual (trained) ones.
-        self.target_encoder.requires_grad_(False)
-        self.target_encoder.load_state_dict(self.encoder.state_dict())
+        if self.is_learner_module:
```
Comment (sven1977) [inline rllib/algorithms/dqn/torch/dqn_rainbow_torch_rl_module.py:40]: Nice!

Comment (sven1977) [reply rllib/algorithms/dqn/torch/dqn_rainbow_torch_rl_module.py:40]: Nit: Should we use super-constrained naming here and either always use either ...?
1) "light"
2) "learner"
but not both terms mixed.

I like the term "light". Or "inference_only".

So - if we chose one of these - we would change the if block to:
```
if not self.is_light_module:
    ...
```

File: rllib/algorithms/dqn/torch/dqn_rainbow_torch_rl_module.py [hunk 2]
```diff
@@ -131,6 +132,11 @@ def _forward_exploration(
     def _forward_train(
         self, batch: Dict[str, TensorType]
     ) -> Dict[str, TensorStructType]:
+        if not self.is_learner_module:
```
Comment (sven1977) [inline rllib/algorithms/dqn/torch/dqn_rainbow_torch_rl_module.py:135]: same nit as above

File: rllib/algorithms/ppo/ppo.py [hunk 3]
```diff
@@ -464,10 +464,11 @@ def _training_step_new_api_stack(self) -> ResultDict:
                     from_worker_or_learner_group=self.learner_group,
                     policies=policies_to_update,
                     global_vars=None,
+                    inference_only=True,
```
Comment (sven1977) [inline rllib/algorithms/ppo/ppo.py:467]: Yeah, I think "inference_only" is the best term. Most descriptive and still short.

Also, I like that the "learner" version is the default. Matches the use-experience for a typical RL setup (you normally train your model in one or another way first, before deploying it to prod).

File: rllib/algorithms/ppo/torch/ppo_torch_rl_module.py [hunk 4]
```diff
@@ -1,4 +1,4 @@
-from typing import Any, Dict
+from typing import Any, Dict, Mapping
```
Comment (sven1977) [inline rllib/algorithms/ppo/torch/ppo_torch_rl_module.py:1]: Can we remove `Mapping` (replace with `Dict`)? I find it confusing. In the end, we work with dicts anyways. Mapping is too generic and doesn't really match here. It's an artifact from earlier times :)

File: rllib/algorithms/ppo/torch/ppo_torch_rl_module.py [hunk 5]
```diff
@@ -17,6 +17,19 @@
 class PPOTorchRLModule(TorchRLModule, PPORLModule):
     framework: str = "torch"
 
+    @override(TorchRLModule)
+    def get_state(self, inference_only: bool = False) -> Mapping[str, Any]:
```
Comment (sven1977) [inline rllib/algorithms/ppo/torch/ppo_torch_rl_module.py:21]: So basically change the signature to Dict (also in the parent method). We don't have to do all Mapping -> Dict in this PR, but we should not introduce new `Mapping` occurrences anymore in new PRs.

File: rllib/algorithms/sac/sac_rl_module.py [hunk 6]
```diff
@@ -68,38 +68,43 @@ def setup(self):
...
-        # TODO (simon): Maybe merging encoders together for target and qf
-        # and keep only the heads differently?
-        self.qf_target_encoder = catalog.build_qf_encoder(framework=self.framework)
-        # If necessary, build also a twin Q encoders.
-        if self.twin_q:
-            self.qf_twin_encoder = catalog.build_qf_encoder(framework=self.framework)
-            self.qf_target_twin_encoder = catalog.build_qf_encoder(
-                framework=self.framework
-            )
+        if self.is_learner_module:
```
Comment (sven1977) [inline rllib/algorithms/sac/sac_rl_module.py:71]: Awesome!

File: rllib/env/multi_agent_env_runner.py [hunk 7]
```diff
@@ -734,7 +734,9 @@ def _make_module(self):
                 },
             )
             ma_rlm_spec: MultiAgentRLModuleSpec = self.config.get_marl_module_spec(
-                policy_dict=policy_dict
+                policy_dict=policy_dict,
```
Comment (sven1977) [inline rllib/env/multi_agent_env_runner.py:737]: Awesome!

File: rllib/evaluation/worker_set.py [hunk 8]
```diff
@@ -448,6 +448,7 @@ def sync_weights(
         to_worker_indices: Optional[List[int]] = None,
         global_vars: Optional[Dict[str, TensorType]] = None,
         timeout_seconds: Optional[int] = 0,
+        inference_only: Optional[bool] = False,
```
Comment (sven1977) [inline rllib/evaluation/worker_set.py:451]: Should we set this to True by default? Or at least add a TODO to switch this?

In the end - on the new API stack - all EnvRunners should only use the `inference_only` version. Also, there might be a time, in which we won't have a local worker anymore and simply use the Learners as a master-copy for all trainable weights.

File: rllib/core/rl_module/rl_module.py [hunk 9]
```diff
@@ -374,9 +374,17 @@ class RLModule(abc.ABC):
     """
 
     framework: str = None
+    is_learner_module: bool = None
 
     def __init__(self, config: RLModuleConfig):
         self.config = config
+        # By default, each module is a learner module and contains all building blocks,
```
Comment (sven1977) [inline rllib/core/rl_module/rl_module.py:381]: nit: Let's use a fixed constant here:

In core/rl_module/__init__.py, add:
```
INFERENCE_ONLY = "_inference_only"
```

Note the underscore to make it harder for users to accidentally override this in their own config dicts.

**Review by sven1977 — APPROVED** [review]
Awesome PR @simonsays1980 . Just some renaming (and other minor) nits!

File: rllib/evaluation/worker_set.py (hunk 8 above)
Comment (simonsays1980) [reply rllib/evaluation/worker_set.py:451]: Good idea. Afaics any synching does only effect from learner to env-runner.

File: rllib/core/rl_module/rl_module.py (hunk 9 above)
Comment (simonsays1980) [reply rllib/core/rl_module/rl_module.py:381]: Love it

**Pull request merged**